src/H5Dmpio.c

```diff
@@ -4952,10 +5500,18 @@ H5D__mpio_collective_filtered_chunk_reinsert(H5D_filtered_collective_io_info_t *
          * they match.
          */
         for (size_t dbg_idx = 0; dbg_idx < chunk_list->num_chunk_infos; dbg_idx++) {
-            if (coll_entry->index_info.chunk_idx == chunk_list->chunk_infos[dbg_idx].index_info.chunk_idx) {
+            bool same_chunk;
+
+            /* Chunks must have the same index and reside in the same dataset */
+            same_chunk = (0 == H5_addr_cmp(coll_entry->index_info.dset_oloc_addr,
+                                           chunk_list->chunk_infos[dbg_idx].index_info.dset_oloc_addr));
+            same_chunk = same_chunk && (coll_entry->index_info.chunk_idx ==
+                                        chunk_list->chunk_infos[dbg_idx].index_info.chunk_idx);
+
+            if (same_chunk) {
                 bool coords_match =
                     !memcmp(scaled_coords, chunk_list->chunk_infos[dbg_idx].chunk_info->scaled,
-                            di->dset->shared->ndims * sizeof(hsize_t));
+                            cached_dset_info->dset->shared->ndims * sizeof(hsize_t));
```

Comment:
If the dataset and index are the same is it still necessary to check the coordinates?

Reply:
Definitely not necessary, though it's very helpful to me in catching implementation bugs. I mostly added this check because of the code and note above where extensible arrays do not act predictably like other chunk index types:

```
             * Extensible arrays where the unlimited dimension is not
             * the slowest-changing dimension "swizzle" the coordinates
             * to move the unlimited dimension value to offset 0. Therefore,
             * we use the "swizzled" down chunks to calculate the "swizzled"
             * scaled coordinates and then we undo the "swizzle" operation.
             *
```

Reply:
Should it be an assertion then?

Reply:
It is just below, though the GitHub view may not be great at showing that